pkg/retester/retester_test.go

```diff
@@ -567,3 +580,195 @@ pr3:
+				HeadRefOID: "pausePR"},
+			policy:         RetesterPolicy{3, 9, &True},
+			expected:       1,
+			expectedString: "Revision pausePR was retested 3 times against base HEAD : pausing",
+		},
+		{
+			name:           "retest PR",
+			cache:          backoffCache{cache: map[string]*pullRequest{}, logger: logger},
+			pr:             tide.PullRequest{HeadRefOID: "retestPR"},
+			policy:         RetesterPolicy{3, 9, &True},
+			expected:       2,
+			expectedString: "Remaining retests: 2 against base HEAD  and 8 for PR HEAD retestPR in total",
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual, actualString := tc.cache.check(tc.pr, tc.baseSha, tc.policy)
+			if diff := cmp.Diff(tc.expected, actual); diff != "" {
+				t.Errorf("%s differs from expected:\n%s", tc.name, diff)
+			}
+			if diff := cmp.Diff(tc.expectedString, actualString); diff != "" {
+				t.Errorf("%s differs from expected:\n%s", tc.name, diff)
+			}
+		})
+	}
+}
+
+var (
+	sevenDays             = 24 * 7 * time.Hour
+	emptyRetestController = &RetestController{}
+)
+
+func TestNewController(t *testing.T) {
+	testCases := []struct {
+		name           string
+		ghClient       githubClient
+		cfg            config.Getter
+		gitClient      git.ClientFactory
+		usesApp        bool
+		cacheFile      string
+		cacheRecordAge time.Duration
+		config         *Config
+		expected       *RetestController
+	}{
+		{
+			name:           "basic",
+			cacheFile:      "basic_case.yaml",
+			cacheRecordAge: sevenDays,
+			expected:       &RetestController{backoff: &backoffCache{file: "testdata/NewController/basic_case.yaml", cacheRecordAge: sevenDays}},
+		},
+		{
+			name:           "empty",
+			cacheFile:      "",
+			cacheRecordAge: sevenDays,
+			expected:       emptyRetestController,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.cacheFile = filepath.Join("testdata", "NewController", tc.cacheFile)
+			actual := NewController(tc.ghClient, tc.cfg, tc.gitClient, tc.usesApp, tc.cacheFile, tc.cacheRecordAge, tc.config)
+			if tc.expected != emptyRetestController {
+				if diff := cmp.Diff(tc.expected.backoff.file, actual.backoff.file); diff != "" {
+					t.Errorf("%s differs from expected:\n%s", tc.name, diff)
+				}
+			}
+		})
+	}
+}
+
+func TestRun(t *testing.T) {
+	config := &Config{Retester: Retester{
+		RetesterPolicy: RetesterPolicy{MaxRetestsForShaAndBase: 3, MaxRetestsForSha: 9}, Oranizations: map[string]Oranization{
+			"org": {RetesterPolicy: RetesterPolicy{Enabled: &True}},
+		},
+	}}
+	ghc := &MyFakeClient{fakegithub.NewFakeClient()}
+	pr123 := github.PullRequest{ID: 1}
+	ghc.PullRequests = map[int]*github.PullRequest{123: &pr123}
+
+	logger := logrus.NewEntry(logrus.StandardLogger())
+
+	testCases := []struct {
+		name       string
+		prowconfig string
+		expected   error
+	}{
+		{
+			name:       "basic",
+			prowconfig: "simple.yaml",
```

Comment:
I am trying to understand the difference between the 2 cases:
Can you show me a line of code for each case that is not covered by the other?

Reply:
You are right we don't need empty case. It's just for test that `empty` works.